File: src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
// #include "example_interfaces/msg/string.hpp"


class NumberPublisherNode : public rclcpp::Node //TODO (Ex: NumberPublisherNode)
{
public:
    NumberPublisherNode() : Node("number_publisher"), number_(2) //TODO (Ex: number_publisher)
    {
        // Declare parameters (and their default values)
        this->declare_parameter("number_to_publish", 2);
        this->declare_parameter("publish_frequency", 1.0);

        number_ = this->get_parameter("number_to_publish").as_int();
        double publish_frequency_ = this->get_parameter("publish_frequency").as_double();

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10); //TODO (Ex: number)
        timer_ = this->create_wall_timer(std::chrono::milliseconds((int) (1000.0 / publish_frequency_)),
                                         std::bind(&NumberPublisherNode::publishNumber, this));

        RCLCPP_INFO(this->get_logger(), "Publisher has been started.");
    }

private:
    void publishNumber() //TODO (Ex: publishNumber)
    {
        auto msg = example_interfaces::msg::Int64(); //TODO: DEFINE FORMAT OF MESSAGE (Ex: String, Int64...)
        msg.data = number_; //TODO: ADD DATA TO MESSAGE
        publisher_->publish(msg);
    }

    std::int64_t number_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64

class NumberPublisherNode(Node): 

    def __init__(self):
        super().__init__("number_publisher")

        # Declare parameters
        self.declare_parameter("number_to_publish", 2) # 2 is the default value for the parameter, invoked when there is not declared a parameter at runtime
        self.declare_parameter("publish_frequency", 1.0)

        self.number_ = self.get_parameter("number_to_publish").value
        self.publish_frequency_ = self.get_parameter("publish_frequency").value

        self.number_publisher_ = self.create_publisher(Int64, "number", 10) # Topic: /number, queue size 10
        self.timer_ = self.create_timer(1.0 / self.publish_frequency_, self.publish_number)

        self.get_logger().info("Number Publisher has been started") 

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_ # Message to be sent to publisher's topic
        self.number_publisher_.publish(msg)



def main(args=None):
    rclpy.init(args=args)
    node = NumberPublisherNode() # TODO
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()